File: python/dnd-character/dnd_character.py
"""D&D Character Generator."""

import random

TRAITS = ('strength', 'dexterity', 'constitution', 'intelligence', 'wisdom', 'charisma')


def modifier(stat: int) -> int:
    """Return the modifier value based on a skill stat."""
    return (stat - 10) // 2


class Character:
    """D&D character generator."""

    # For pylint
    constitution: int

    def __init__(self):
        """Initialize the character."""
        for trait in TRAITS:
            setattr(self, trait, self.ability())

        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self) -> int:
        """Roll an ability stat: 4d6 and discard the lowest die."""
        dice = sorted(random.randint(1, 6) for _ in range(4))
        return sum(dice[1:])

File: go/dnd-character/dnd_character.go
package dndcharacter

import "math/rand"

type Character struct {
	Strength     int
	Dexterity    int
	Constitution int
	Intelligence int
	Wisdom       int
	Charisma     int
	Hitpoints    int
}

// Modifier calculates the ability modifier for a given ability score
func Modifier(score int) int {
	if score < 10 {
		score--
	}
	return (score - 10) / 2
}

// Ability uses randomness to generate the score for an ability
func Ability() int {
	var total int
	smallest := 7
	for range 4 {
		roll := rand.Intn(6) + 1
		if roll < smallest {
			smallest = roll
		}
		total += roll
	}
	return total - smallest
}

// GenerateCharacter creates a new Character with random scores for abilities
func GenerateCharacter() Character {
	con := Ability()
	return Character{
		Ability(),
		Ability(),
		con,
		Ability(),
		Ability(),
		Ability(),
		10 + Modifier(con),
	}
}
